pywt/src/convolution.c.src

```diff
@@ -7,768 +7,397 @@
  * #type = double, float#
  */
 
-int @type@_downsampling_convolution_periodization(const @type@* input, const_index_t N,
-                             const @type@* filter, const_index_t F,
-                             @type@* output, const_index_t step)
+/* This file contains several functions for computing the convolution of a
+ * signal with a filter. The general scheme is:
+ *   output[o] = sum(filter[j] * input[i-j] for j = [0..F) and i = [0..N))
+ * where 'o', 'i' and 'j' may progress at different rates.
+ *
+ * Most of the code deals with different edge extension modes. Values are
+ * computed on-demand, in four steps:
+ * 1. Filter extends past signal on the left.
+ * 2. Filter completely contained within signal (no extension).
+ * 3. Filter extends past signal on both sides (only if F > N).
+ * 4. Filter extends past signal on the right.
+ *
+ * MODE_PERIODIZATION produces different output lengths to other modes, so is
+ * implemented as a separate function for each case.
+ *
+ * See 'common.h' for descriptions of the extension modes.
+ */
 
+int @type@_downsampling_convolution_periodization(const @type@ * const restrict input, const size_t N,
+                                                  const @type@ * const restrict filter, const size_t F,
+                                                  @type@ * const restrict output, const size_t step)
 {
-
-    index_t i, j, k, F_2, corr;
-    index_t start;
-    @type@ sum;
-    @type@* ptr_w = output;
-
-    i = step-1;  /* first element taken from input is input[step-1] */
-    start = F_2 = F/2;  /* extending by (F-2)/2 elements */
-    corr = 0;
-
-    for(i=start; i < F; i+=step){
-        sum = 0;
-        for(j = 0; j < i+1-corr; ++j)    /* overlapping */
-            sum += filter[j]*input[i-j-corr];
-
-        if(N%2){
-            if(F-j){ /* if something to extend */
+    size_t i = F/2, o = 0;
+    const size_t padding = N % 2; // FIXME: Should be N%step for larger steps?
```

Comment:
I'd just like to draw attention to this. My thought is this should be `N%step` to give the intended behaviour (i.e. pad signal lengths out to multiples of the downsampling step), but I just want to confirm this.